micrometer-core/src/test/java/io/micrometer/core/instrument/internal/DefaultLongTaskTimerTest.java

```diff
@@ -0,0 +1,115 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class DefaultLongTaskTimerTest {
+
+    private static final InstanceOfAssertFactory<SampleImpl, AbstractComparableAssert<?, SampleImpl>> SAMPLE_IMPL_ASSERT = new InstanceOfAssertFactory<>(
+            SampleImpl.class, Assertions::assertThat);
+
+    @Test
+    void sampleTimestampCollision() {
+        final MockClock clock = new MockClock();
+        MeterRegistry registry = new SimpleMeterRegistry(SimpleConfig.DEFAULT, clock);
+        LongTaskTimer t = LongTaskTimer.builder("my.timer").register(registry);
+
+        final LongTaskTimer.Sample sample1 = t.start();
+        final LongTaskTimer.Sample sample2 = t.start();
+        assertThat(sample1).isInstanceOf(SampleImpl.class).isNotInstanceOf(SampleImplCounted.class);
+        assertThat(sample2).isInstanceOf(SampleImplCounted.class).isNotSameAs(sample1);
+        final SampleImpl sampleImpl1 = (SampleImpl) sample1;
+        final SampleImpl sampleImpl2 = (SampleImpl) sample2;
+
+        clock.addSeconds(1);
+
+        assertThat(sample1.duration(TimeUnit.SECONDS)).isEqualTo(1);
+        assertThat(sample2.duration(TimeUnit.SECONDS)).isEqualTo(1);
+
+        final LongTaskTimer.Sample sample3 = t.start();
+        final LongTaskTimer.Sample sample4 = t.start();
+        assertThat(sample3).isInstanceOf(SampleImpl.class).isNotInstanceOf(SampleImplCounted.class);
+        assertThat(sample4).isInstanceOf(SampleImplCounted.class);
+        final SampleImpl sampleImpl3 = (SampleImpl) sample3;
+        final SampleImpl sampleImpl4 = (SampleImpl) sample4;
+
+        assertThat(t.activeTasks()).isEqualTo(4);
+
+        assertThat(sampleImpl4).isEqualByComparingTo(sampleImpl4)
+            .isGreaterThan(sampleImpl3)
+            .isGreaterThan(sampleImpl2)
+            .isGreaterThan(sampleImpl1);
+        assertThat(sampleImpl3).isEqualByComparingTo(sampleImpl3)
+            .isLessThan(sampleImpl4)
+            .isGreaterThan(sampleImpl2)
+            .isGreaterThan(sampleImpl1);
+        assertThat(sampleImpl2).isEqualByComparingTo(sampleImpl2)
+            .isLessThan(sampleImpl4)
+            .isLessThan(sampleImpl3)
+            .isGreaterThan(sampleImpl1);
+        assertThat(sampleImpl1).isEqualByComparingTo(sampleImpl1)
+            .isLessThan(sampleImpl4)
+            .isLessThan(sampleImpl3)
+            .isLessThan(sampleImpl2);
+
+        assertThat(sample2.stop()).isEqualTo(TimeUnit.SECONDS.toNanos(1));
+        assertThat(t.activeTasks()).isEqualTo(3);
+        assertThat(sample3.stop()).isEqualTo(TimeUnit.SECONDS.toNanos(0));
+        assertThat(t.activeTasks()).isEqualTo(2);
+        assertThat(sample4.stop()).isEqualTo(TimeUnit.SECONDS.toNanos(0));
+        assertThat(t.activeTasks()).isEqualTo(1);
+        assertThat(sample1.stop()).isEqualTo(TimeUnit.SECONDS.toNanos(1));
+        assertThat(t.activeTasks()).isEqualTo(0);
+    }
+
+    @Test
+    void counterJumpsZeroAndWraps() {
```

Comment:
Can we split this into two?
1. `counterShouldSurviveOverflow`
2. `counterShouldJumpZero`